extensions/wikia/CreateNewWiki/CreateNewWikiController.class.php

```diff
@@ -228,132 +241,137 @@ public function CreateWiki() {
-				$this->status = 'wikilimit';
-				$this->statusMsg = wfMsgExt('cnw-error-wiki-limit', array( 'parsemag' ), self::DAILY_USER_LIMIT);
-				$this->statusHeader = wfMsg('cnw-error-wiki-limit-header');
-				wfProfileOut(__METHOD__);
-				return;
-			}
+		// check if user has confirmed e-mail
+		if ( !$wgUser->isEmailConfirmed() ) {
+			$this->warning(__METHOD__ . ": user's email not confirmed" );
+			$this->response->setCode( 403 );
+			$this->response->setVal( self::STATUS_FIELD, self::STATUS_ERROR );
+			$this->response->setVal( self::STATUS_MSG_FIELD, wfMessage( 'cnw-error-unconfirmed-email' )->parse() );
+			$this->response->setVal( self::STATUS_HEADER_FIELD, wfMessage( 'cnw-error-unconfirmed-email-header' )->text() );
+			wfProfileOut(__METHOD__);
+			return;
+		}
 
-			$categories = isset($params['wCategories']) ? $params['wCategories'] : array();
+		// check if user is blocked
+		if ( $wgUser->isBlocked() ) {
+			$this->warning(__METHOD__ . ": user is blocked" );
+			$this->response->setCode( 403 );
+			$this->response->setVal( self::STATUS_FIELD, self::STATUS_ERROR );
+			$this->response->setVal( self::STATUS_MSG_FIELD, wfMessage( 'cnw-error-blocked', $wgUser->blockedBy(), $wgUser->blockedFor(), $wgUser->getBlockId() )->parse() );
+			$this->response->setVal( self::STATUS_HEADER_FIELD, wfMessage( 'cnw-error-blocked-header' )->text() );
+			wfProfileOut(__METHOD__);
+			return;
+		}
 
-			$createWiki = new CreateWiki($params['wName'], $params['wDomain'], $params['wLanguage'], $params['wVertical'], $categories);
+		// check if user is a tor node
+		if ( class_exists( 'TorBlock' ) && TorBlock::isExitNode() ) {
+			$this->warning(__METHOD__ . ": user is blocked (TOR detected)" );
+			$this->response->setCode( 403 );
+			$this->response->setVal( self::STATUS_FIELD, self::STATUS_ERROR );
+			$this->response->setVal( self::STATUS_MSG_FIELD, wfMessage( 'cnw-error-torblock' )->text() );
+			$this->response->setVal( self::STATUS_HEADER_FIELD, wfMessage( 'cnw-error-blocked-header' )->text() );
+			wfProfileOut(__METHOD__);
+			return;
+		}
 
-			try {
-				$createWiki->create();
-			}
-			catch(Exception $ex) {
-				$error_code = $ex->getCode();
+		// check if user created more wikis than we allow per day
+		$numWikis = $this->countCreatedWikis($wgUser->getId());
+		if($numWikis >= self::DAILY_USER_LIMIT && $wgUser->isPingLimitable() && !$wgUser->isAllowed( 'createwikilimitsexempt' ) ) {
+			$this->warning(__METHOD__ . ": user reached daily creation count limit" );
+			$this->response->setCode( 429 );
+			$this->response->setVal( self::STATUS_FIELD, self::STATUS_CREATION_LIMIT );
+			$this->response->setVal( self::STATUS_MSG_FIELD, wfMessage( 'cnw-error-wiki-limit', self::DAILY_USER_LIMIT )->parse() );
+			$this->response->setVal( self::STATUS_HEADER_FIELD, wfMessage( 'cnw-error-wiki-limit-header' )->text());
+			wfProfileOut(__METHOD__);
+			return;
+		}
```

Comment:
can we move each of this 5 blocks to a separate function, each for one case ie. when user not logged, when has not confirmed email...